autopilot/betweenness_centrality.go

```diff
@@ -0,0 +1,177 @@
+package autopilot
+
+// stack is a simple int stack to help with readability of Brandes'
+// betweenness centrality implementation below.
+type stack struct {
+	stack []int
+}
+
+func (s *stack) Push(v int) {
+	s.stack = append(s.stack, v)
+}
+
+func (s *stack) Top() int {
+	return s.stack[len(s.stack)-1]
+}
+
+func (s *stack) Pop() {
+	s.stack = s.stack[:len(s.stack)-1]
+}
+
+func (s *stack) Empty() bool {
+	return len(s.stack) == 0
+}
+
+// queue is a simple int queue to help with readability of Brandes'
+// betweenness centrality implementation below.
+type queue struct {
+	queue []int
+}
+
+func (q *queue) Push(v int) {
+	q.queue = append(q.queue, v)
+}
+
+func (q *queue) Front() int {
+	return q.queue[0]
+}
+
+func (q *queue) Pop() {
+	q.queue = q.queue[1:]
+}
+
+func (q *queue) Empty() bool {
+	return len(q.queue) == 0
+}
+
+// BetweennessCentrality is a NodeMetric that calculates node betweenness
+// centrality using Brandes' algorithm. Betweenness centrality for each node
+// is the number of shortest paths passing trough that node, not counting
+// shortest paths starting or ending at that node. This is a useful metric
+// to measure control of individual nodes over the whole network.
+type BetweennessCentrality struct {
+	centrality map[NodeID]float64
+}
+
+// NewBetweennessCentralityMetric creates a new BetweennessCentrality instance.
+func NewBetweennessCentralityMetric() *BetweennessCentrality {
+	return &BetweennessCentrality{}
+}
+
+// Name returns the name of the metric.
+func (bc *BetweennessCentrality) Name() string {
+	return "betweeness_centrality"
+}
+
+// betweennessCentrality is the core of Brandes' algorithm.
+// We first calculate the shortest paths from the start node s to all other
+// nodes with BFS, then update the betweenness centrality values by using
+// Brandes' dependency trick.
+// For detailed explanation please read:
+// https://www.cl.cam.ac.uk/teaching/1617/MLRD/handbook/brandes.html
+func betweennessCentrality(g *SimpleGraph, s int, centrality []float64) {
+	// Pred[w] is the list of nodes that immediately precede w on a
+	// shortest path from s to t for each node t.
+	Pred := make([][]int, len(g.Nodes))
```

Comment:
Variable names should be all lower case, applies below as well.

Reply:
Yeah, I kept variable names like this to make it easier to follow the linked paper. Can rename if needed though.